R/textstat-methods.R

```diff
@@ -1,71 +1,53 @@
-#' \code{"glob"}, \code{"regex"} or \code{"fixed"} patterns using this method.
-#' @param x a \code{textstat} object
+#' Use filters to find rows of the return objects from many \code{textstat_*()}
+#' functions where conditions are true, including matches to features using
+#' \code{"glob"}, \code{"regex"} or \code{"fixed"} patterns.
+#' @param .data a \code{textstat} object returned from one of the applicable
+#' \code{textstat_*} functions in \pkg{quanteda}
+#' @param ... filter conditions passed to \pkg{dplyr} \code{\link[dplyr]{filter}}
 #' @inheritParams pattern
-#' @param selection whether to \code{"keep"} or \code{"remove"} the rows that
-#'   match the pattern
 #' @inheritParams valuetype
 #' @param case_insensitive ignore case when matching, if \code{TRUE}
+#' @seealso  \code{\link{textstat_collocations}},
+#' \code{\link{textstat_keyness}}, \code{\link{textstat_frequency}}
 #' @keywords textstat internal
 #' @export
 #' @examples
 #' period <- ifelse(docvars(data_corpus_inaugural, "Year") < 1945, "pre-war", "post-war")
 #' mydfm <- dfm(data_corpus_inaugural, groups = period)
 #' keyness <- textstat_keyness(mydfm)
-#' textstat_select(keyness, 'america*')
-#' 
-textstat_select <- function(x, 
+#' filter(keyness, pattern = "america*")
+#' filter(keyness, p < .00001, pattern = "america*") 
+#' filter(keyness, p < .00001) %>% head()
+filter.textstat <- function(.data, ...,
                             pattern = NULL, 
-                            selection = c("keep", "remove"), 
                             valuetype = c("glob", "regex", "fixed"),
                             case_insensitive = TRUE) {
-    UseMethod("textstat_select")
-}
-
-#' @export
-textstat_select.default <- function(x, pattern = NULL, 
-                                  selection = c("keep", "remove"), 
-                                  valuetype = c("glob", "regex", "fixed"),
-                                  case_insensitive = TRUE) {
-    stop(friendly_class_undefined_message(class(x), "textstat_select"))
-}
-
-#' @export
-textstat_select.textstat <- function(x, pattern = NULL,
-                                     selection = c("keep", "remove"), 
-                                     valuetype = c("glob", "regex", "fixed"), 
-                                     case_insensitive = TRUE) {
     
-    if (is.null(pattern)) return(x)
-    selection <- match.arg(selection)
-    valuetype <- match.arg(valuetype)
-    attrs <- attributes(x)
+    attrs <- attributes(.data)
     
-    id <- unlist(regex2id(pattern, x[[1]], valuetype, case_insensitive))
-    if (selection == 'keep') {
-        x <- x[id,]
-    } else {
-        x <- x[id * -1,]
+    # call dplyr filter, if ... arguments are supplied
+    if (length(list(substitute(...)))) {
+        if (!requireNamespace("dplyr", quietly = TRUE)) 
+            stop("You must install the dplyr package to use filter.textstat")
+        .data <- as.data.frame(dplyr::filter(dplyr::tbl_df(.data), ...))
```

Comment:
Why do you coerce back and forth to a tibble here?

Reply:
For some reason I had trouble accessing `dplyr::filter.data.frame`. But this solves it:
```r
dplyr:::filter.data.frame(.data, ...)
```